### PR TITLE
Feature Addition: Group Syncing

### PR DESCRIPTION
**Pull request by sillywilly42**
Based on implementation suggestion by Justin McWilliams:
https://groups.google.com/d/msg/simian-discuss/2CNN5EvaRWs/MpeetxHkDUEJ

If this pull request gets merged I'll take a stab at updating the API documentation to include an example client script.

This change creates a new Manifest Modification type: group. There's a group model which is just a list of owners (strings). I created a new API for creating, updating, viewing, and deleting groups. And the web interface has been updated to include groups. A lot of the code is just duplication of the Tag code since Tags are to uuids what Groups are to owners.

Most changes support this new feature, however the following additional changes were also made:
* src/simian/mac/common/auth.py - added scope string. I could not make oauth with a service account work without this. I think something has maybe changed on the appengine backend?
* src/simian/mac/client/flight_common.py - the client parses the output of `last` to work out the most frequent user, on new machines that could be ‘reboot’ under certain conditions.

### CODE REVIEW

File: src/simian/mac/admin/groups.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+import urllib
+
+from google.appengine.ext import db
+
+from simian.mac import admin
+from simian.mac import models
+from simian.mac.common import auth
+
+
+class Groups(admin.AdminHandler):
```
Comment (maximermilov) [inline src/simian/mac/admin/groups.py]: Please write tests for new feature.

File: src/simian/mac/admin/host.py [hunk 2]
```diff
@@ -139,6 +141,15 @@ def _DisplayHost(self, computer, self_report):
           tags[tag] = False
       tags = json.dumps(tags, sort_keys=True)
 
+      # Generate groups data.
```
Comment (maximermilov) [inline src/simian/mac/admin/host.py]: tests ;)

File: src/simian/mac/common/applesus.py [hunk 3]
```diff
@@ -34,7 +34,7 @@
 
 
 OS_VERSIONS = frozenset(
-    ['10.7', '10.8', '10.9', '10.10', '10.11'])
```
Comment (maximermilov) [inline src/simian/mac/common/applesus.py]: Please revert unrelated changes.

File: src/simian/auth/tlslite_bridge.py [hunk 4]
```diff
@@ -30,7 +30,7 @@
   # we don't use this api, so safely stub out fcntl for appengine
   if (os.environ.get('SERVER_SOFTWARE', '').startswith('Google App Engine') or
       os.environ.get('SERVER_SOFTWARE', '').startswith('Development')):
-    logging.warning('stub out fcntl')
+    logging.info('stub out fcntl')
```
Comment (maximermilov) [inline src/simian/auth/tlslite_bridge.py]: > otherwise the logs are filled with this warning.

which logs?

I think, it will be easier to have unrelated small changes as separate pull request.

File: src/simian/mac/admin/__init__.py [hunk 5]
```diff
@@ -221,8 +222,7 @@ def Render(self, template_path, values, write_to_response=True):
     path = os.path.join(
         os.path.dirname(__file__), 'templates', template_path)
 
-    if not settings.DEV_APPSERVER:
-      values['static_path'] = 'myapp/%s' % os.getenv('CURRENT_VERSION_ID')
+    values['static_path'] = 'myapp/%s' % os.getenv('CURRENT_VERSION_ID')
```
Comment (maximermilov) [inline src/simian/mac/admin/__init__.py]: why?

File: src/simian/mac/api/groups.py [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+import webapp2
+
+from simian import settings
+from simian.mac import models
+from simian.mac.common import auth
+
+API_INFO_KEY = settings.API_INFO_KEY
+
+
+class GroupHandler(webapp2.RequestHandler):
```
Comment (maximermilov) [inline src/simian/mac/api/groups.py]: tests ;)

File: src/simian/mac/admin/__init__.py (hunk 5 above)
Comment (sillywilly42) [inline src/simian/mac/admin/__init__.py]: With this conditional in place, the dev_appserver doesn't seem to serve up CSS etc. If I remove the conditional everything is fine.
I'm not the only one: https://groups.google.com/d/msg/simian-discuss/hRE2ieA_a14/9YoPqy6qeXAJ

File: src/simian/mac/admin/groups.py (hunk 1 above)
Comment (sillywilly42) [inline src/simian/mac/admin/groups.py]: Done!

File: src/simian/mac/admin/host.py (hunk 2 above)
Comment (sillywilly42) [inline src/simian/mac/admin/host.py]: There aren't existing tests for this module, perhaps we could forgo this one?

File: src/simian/mac/api/groups.py (hunk 6 above)
Comment (sillywilly42) [inline src/simian/mac/api/groups.py]: Done!

File: src/simian/auth/tlslite_bridge.py (hunk 4 above)
Comment (sillywilly42) [inline src/simian/auth/tlslite_bridge.py]: In the AppEngine console logs - I like to monitor at warning level...
Reverted for now.

File: src/tests/simian/mac/admin/groups_test.py [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+    request = None
+    response = mock.MagicMock()
+    return groups.Groups(request, response)
+
+  def GetTestClassModule(self):
+    return groups
+
+  def testGet(self, _):
+    """Test get()."""
+    self.mox.StubOutWithMock(self.c, 'IsAdminUser')
```
Comment (maximermilov) [inline src/tests/simian/mac/admin/groups_test.py]: mox is deprecated and should not be used in new tests.

New tests should use testbed and mock.
good example acl_groups_test.py

File: src/simian/mac/admin/host.py (hunk 2 above)
Comment (maximermilov) [inline src/simian/mac/admin/host.py]: hmm, ok

File: src/tests/simian/mac/admin/groups_test.py (hunk 7 above)
Comment (sillywilly42) [inline src/tests/simian/mac/admin/groups_test.py]: I've rewritten tests not to use mox.

File: src/simian/mac/api/groups.py [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+
+    group = self.request.get('group')
+    if group:
+      existing_group = models.Group.get_by_key_name(group)
+      if existing_group:
+        self.response.out.write(json.dumps(existing_group.owners))
+    else:
+      self.response.out.write(json.dumps(models.Group.GetAllGroupNames()))
+
+  def post(self):
```
Comment (maximermilov) [inline src/simian/mac/api/groups.py]: vulnerable to xsrf attack.
Seems, this class duplicate functionality from admin/groups.py

Comment (sillywilly42) [inline src/simian/mac/api/groups.py]: This class is for the stateless API - which is the motivation behind this PR. I added the web interface bits exposing the groups feature as a courtesy to other Simian users.

I can't add a CSRF token because then this wouldn't be stateless. Instead I pass the existing API_INFO_KEY as a header (X-Simian-API-Info-Key) for all HTTP methods.

File: src/simian/mac/api/groups.py [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+      self.response.set_status(httplib.UNAUTHORIZED)
+      return
+
+  def get(self):
+    """List groups, or list a group's memberhsip.
+
+    Send optional 'group' parameter to list membership of a specified group.
+    Otherwise a list of groups is returned.
+    """
+    self._check_api_key()
```
Comment (maximermilov) [inline src/simian/mac/api/groups.py]: Why there is auth check for all other methods but not this one?

Comment (sillywilly42) [inline src/simian/mac/api/groups.py]: I originally thought the get handler would work the same way it does for the packages API.
I have now added the oauth check to bring it inline with the others.

**Comment by maximermilov** [comment]
Thank you.
merged with minor fixes.
( https://github.com/google/simian/commit/7a67c521280ef81ee0fcf689b7566df02639b4fb )
